prov/gni/src/gnix_msg.c

```diff
@@ -1651,3 +2404,441 @@ ssize_t _gnix_send(struct gnix_fid_ep *ep, uint64_t loc_addr, size_t len,
 	}
 	return ret;
 }
+
+ssize_t _gnix_recvv(struct gnix_fid_ep *ep, const struct iovec *iov,
+		    void **desc, size_t count, uint64_t src_addr, void *context,
+		    uint64_t flags, uint64_t ignore, uint64_t tag)
+{
+	int i, ret = FI_SUCCESS;
+	size_t cum_len = 0;
+	struct gnix_fab_req *req = NULL;
+	struct gnix_address gnix_addr;
+	fastlock_t *queue_lock = NULL;
+	struct gnix_tag_storage *posted_queue = NULL;
+	struct gnix_tag_storage *unexp_queue = NULL;
+	uint64_t match_flags;
+	int tagged = flags & FI_TAGGED;
+	struct fid_mr *auto_mr;
+
+	if (!ep->recv_cq) {
+		return -FI_ENOCQ;
+	}
+
+	if (!tagged) {
+		if (!ep->ep_ops.msg_send_allowed)
+			return -FI_EOPNOTSUPP;
+
+		tag = 0;
+		ignore = ~0;
+	} else {
+		if (!ep->ep_ops.tagged_send_allowed)
+			return -FI_EOPNOTSUPP;
+	}
+
+	match_flags = flags & (FI_CLAIM | FI_DISCARD | FI_PEEK);
+
+	/*
+	 * Lookup the gni addr in the av_table or av_hashtable.
+	 * If the gni addr doesn't exist the addr is FI_ADDR_UNSPEC,
+	 * meaning this remote node wants to receive from all senders?
+	 */
+	ret = __gnix_msg_addr_lookup(ep, src_addr, &gnix_addr);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	/* calculate cumulative size of the iovec buf lens */
+	for (i = 0; i < count; i++) {
+		cum_len += iov[i].iov_len;
+	}
+
+	/*
+	 * Initialize the tag storage objects.
```

Comment:
Nit pick: we don't initialize tag storage here.

I think you can remove the question marks from the comments, I think the statements are correct.  The only vague thing is that source address matching is not on by default but is possible by using FI_DIRECTED_RECV.